README.md
# Logstash TCP stream for Bunyan

[![build status](https://secure.travis-ci.org/chris-rock/bunyan-logstash-tcp.png)](http://travis-ci.org/chris-rock/bunyan-logstash-tcp)

A tcp logger for [Logstash](http://logstash.net/docs/1.4.2/inputs/tcp)

## Configuration options

<table>
  <tr>
    <th>level</th>
    <td>string</td>
    <td><code>info</code></td>
  </tr>
  <tr>
    <th>server</th>
    <td>string</td>
    <td><code>os.hostname()</code></td>
  </tr>
  <tr>
    <th>host</th>
    <td>string</td>
    <td><code>"127.0.0.1"</code></td>
  </tr>
  <tr>
    <th>port</th>
    <td>number</td>
    <td><code>9999</code></td>
  </tr>
  <tr>
    <th>application</th>
    <td>string</td>
    <td><code>process.title</code></td>
  </tr>
  <tr>
    <th>pid</th>
    <td>string</td>
    <td><code>process.pid</code></td>
  </tr>
  <tr>
    <th>tags</th>
    <td>array|string[]</td>
    <td><code>["bunyan"]</code></td>
  </tr>
</table>

## Adding the bunyan-logstash stream to Bunyan

```
var log = bunyan.createLogger({
  streams: [
    {
      type: "raw",
      stream: require('bunyan-logstash-tcp').createStream({
        host: '127.0.0.1',
        port: 9908
      })
    }
  ]
});
```

## Example

```javascript
"use strict";

var bunyan = require('bunyan'),
    bunyantcp = require('bunyan-logstash-tcp');

var log = bunyan.createLogger({
    name: 'example',
    streams: [{
        level: 'debug',
        stream: process.stdout
    },{
        level: 'debug',
        type: "raw",
        stream: bunyantcp.createStream({
            host: '127.0.0.1',
            port: 9998
        })
    }],
    level: 'debug'
});

log.debug('test');
log.error('error test');
```

## Logstash Configuration

Configuration for [Logstash 1.3.3+](http://logstash.net/docs/1.4.2/inputs/tcp):

```javascript
input {
  // config for bunyan udp
  udp {
      'port' => "9999"
  }
  // config for bunyan tcp
  tcp {
      'port' => "9998"
  }
}
```

## Try with logstash locally

 - Download logstash from http://logstash.net/
 - Unpack it (tar -zxf logstash-1.4.2.tar.gz)
 - Create a test logstash configuration `logstash.conf`

```code
input {
  stdin { 
    type => "stdin-type"
  }
  udp {
    port => "9999"
  }
  tcp {
    port => "9998"
  }
}
output { 
  stdout {}
}
```

 - Run `bin/logstash agent -f logstash.conf
 - Run `node example/log.js`

## Credits

This module is heavily based on [bunyan-logstash](https://github.com/sheknows/bunyan-logstash) and re-uses parts of [winston-logstash](https://github.com/jaakkos/winston-logstash/blob/master/lib/winston-logstash.js).

Thanks to

- [sheknows](https://github.com/sheknows)
- [jaakkos](https://github.com/jaakkos) 

for their amazing work

## License

MIT


CHANGELOG.md
# CHANGELOG

## 1.0.0

- Drop Node 0.x support

## 0.3.4

- #13 Emit connection events 

## 0.3.3
 
 - add unref (@see http://nodejs.org/api/net.html#net_server_unref) by (liukun)

## 0.3.2

 - add bunyan as peer dependency

## 0.3.1

 - #7 bugfix additional fields merge (jcharles-d2l)

## 0.3.0

 - #6 introduce CBuffer for log cache (futurechan)
 - add more parameters (futurechan)


## 0.2.0 

 - #2 use new logstash format (analytically)

## 0.1.1

 - initial release

test/eventEmitter.js
"use strict";

var bunyan = require('bunyan'),
    bunyantcp = require('../lib/logstash');

var net = require('net'),
    should = require('should');

suite('EventEmitter', function() {

    var server;
    var serverPort;

    suiteSetup(function(done){
        server = net.createServer();
        server.listen(function() {
            serverPort = server.address().port;
            done();
        });
    });
 
    test('should emit a connect event', function(done){
        bunyan.createLogger({ name: 'example', streams: [ { level: 'debug', type: 'raw', stream: bunyantcp.createStream({
            host: '127.0.0.1',
            port: serverPort
        }).on('connect', done)}]});
    });

    test('should emit a close event', function(done){
        var tcps = bunyantcp.createStream({
            host: '127.0.0.1',
            port: serverPort
        });
        bunyan.createLogger({ name: 'example', streams: [ { level: 'debug', type: 'raw', stream: tcps }]});
        tcps.on('connect', function() {
            tcps.socket.end();
        }).on('close', done)
    });

    test('should emit an error event', function(done){
        bunyan.createLogger({ name: 'example', streams: [ { level: 'debug', type: 'raw', stream: bunyantcp.createStream({
            host: '127.0.0.1',
            port: 12
        }).on('error', function(err) { 
            err.should.be.an.instanceof(Error);
            done(); 
        })}]});
    });
});

src/test/logstash.test.js
const os = require('os');
const fs = require('fs');
const net = require('net');
const sinon = require('sinon');
const tls = require('tls');

const CBuffer = require('CBuffer');
const { expect } = require('chai');

const { createStream } = require('../lib/logstash');

const EventEmitter = require('events').EventEmitter;

class MockSocket {
  constructor() {
    this.listeners = {};
    this.unrefCalled = false;
    this.destroyCalled = false;
    this.encoding = null;
    this.host = null;
    this.port = null;
    this.content = '';
  }

  connect(host, port, callback) {
    this.host = host;
    this.port = port;
    setTimeout(callback);
  }

  on(event, callback) {
    this.listeners[event] = this.listeners[event] || [];
    this.listeners[event].push(callback);
  }

  unref() {
    this.unrefCalled = true;
  }

  destroy() {
    this.destroyCalled = true;
  }

  setEncoding(encoding) {
    this.encoding = encoding;
  }

  write(text) {
    this.content += text;
  }

  dispatchEvent(event, ...params) {
    (this.listeners[event] || []).forEach(callback => callback(...params));
  }
}

describe('logstash', () => {
  const sandbox = sinon.sandbox.create();

  beforeEach(() => {
    sandbox.stub(net, 'Socket').callsFake(() => new MockSocket());
  });

  afterEach(() => {
    sandbox.restore();
  });

  describe('createStream', () => {
    it('Should call EventEmitter.call()', () => {
      const callSpy = sandbox.spy(EventEmitter, 'call');

      const stream = createStream();

      expect(callSpy.calledWith(stream));
    });

    it('Should call apply default options when no options are provided', () => {
      const stream = createStream();

      expect(stream).to.have.property('name', 'bunyan');
      expect(stream).to.have.property('level', 'info');
      expect(stream).to.have.property('server', os.hostname());
      expect(stream).to.have.property('host', '127.0.0.1');
      expect(stream).to.have.property('port', 9999);
      expect(stream).to.have.property('application', process.title);
      expect(stream).to.have.property('pid', process.pid);
      expect(stream).to.have.property('tags').eql(['bunyan']);
      expect(stream).to.have.property('type', undefined);
      expect(stream).to.have.property('client', null);
      expect(stream).to.have.property('ssl_enable', false);
      expect(stream).to.have.property('ssl_key', '');
      expect(stream).to.have.property('ssl_cert', '');
      expect(stream).to.have.property('ca', '');
      expect(stream).to.have.property('ssl_passphrase', '');
      expect(stream).to.have.property('cbuffer_size', 10);
      expect(stream).to.have.property('log_queue').instanceOf(CBuffer).deep.property('data.length', 10);
      expect(stream).to.have.property('connected', false);

      expect(stream).to.have.property('socket').instanceOf(MockSocket);
      expect(stream).to.have.property('retries', 0);
      expect(stream).to.have.property('max_connect_retries', 4);
      expect(stream).to.have.property('retry_interval', 100);
    });

    it('Should use the level option value', () => {
      const stream = createStream({ level: 'debug' });

      expect(stream).to.have.property('level', 'debug');
    });

    it('Should use the server option value', () => {
      const stream = createStream({ server: 'myServer.com' });

      expect(stream).to.have.property('server', 'myServer.com');
    });

    it('Should use the host option value', () => {
      const stream = createStream({ host: 'host1' });

      expect(stream).to.have.property('host', 'host1');
    });

    it('Should use the port option value', () => {
      const stream = createStream({ port: 12345 });

      expect(stream).to.have.property('port', 12345);
    });

    it('Should use the appName option value', () => {
      const stream = createStream({ appName: 'my app' });

      expect(stream).to.have.property('application', 'my app');
    });

    it('Should use the pid option value', () => {
      const stream = createStream({ pid: 123456 });

      expect(stream).to.have.property('pid', 123456);
    });

    it('Should use the tags option value', () => {
      const stream = createStream({ tags: ['tag1', 'tag2'] });

      expect(stream).to.have.property('tags').eql(['tag1', 'tag2']);
    });

    it('Should use the type option value', () => {
      const stream = createStream({ type: 'alpha' });

      expect(stream).to.have.property('type', 'alpha');
    });

    it('Should use the ssl_enable option value', () => {
      sandbox.stub(tls, 'connect').callsFake(() => new MockSocket());

      const stream = createStream({ ssl_enable: true });

      expect(stream).to.have.property('ssl_enable', true);
    });

    it('Should use the ssl_key option value', () => {
      const stream = createStream({ ssl_key: 'myKey' });

      expect(stream).to.have.property('ssl_key', 'myKey');
    });

    it('Should use the ssl_cert option value', () => {
      const stream = createStream({ ssl_cert: 'myCert' });

      expect(stream).to.have.property('ssl_cert', 'myCert');
    });

    it('Should use the ca option value', () => {
      const stream = createStream({ ca: 'caaaa' });

      expect(stream).to.have.property('ca', 'caaaa');
    });

    it('Should use the ssl_passphrase option value', () => {
      const stream = createStream({ ssl_passphrase: 'pass' });

      expect(stream).to.have.property('ssl_passphrase', 'pass');
    });

    it('Should use the cbuffer_size option value', () => {
      const stream = createStream({ cbuffer_size: 14 });

      expect(stream).to.have.property('cbuffer_size', 14);
      expect(stream).to.have.property('log_queue').instanceOf(CBuffer).deep.property('data.length', 14);
    });

    it('Should use the numeric max_connect_retries option value', () => {
      const stream = createStream({ max_connect_retries: 12 });

      expect(stream).to.have.property('max_connect_retries', 12);
    });

    it('Should ignore the non numeric max_connect_retries option value', () => {
      const stream = createStream({ max_connect_retries: '12' });

      expect(stream).to.have.property('max_connect_retries', 4);
    });

    it('Should use the retry_interval option value', () => {
      const stream = createStream({ retry_interval: 200 });

      expect(stream).to.have.property('retry_interval', 200);
    });

    it('Should inherit EventEmitter', () => {
      const stream = createStream();
      expect(stream).to.be.instanceOf(EventEmitter);
    });
  });

  describe('LogstashStream', () => {
    describe('write', () => {
      it('Should call the send method', () => {
        const stream = createStream({
          tags: ['tag1', 'tag2'],
          server: 'myServer.com',
          appName: 'myApp',
          pid: 12345
        });

        const sendStub = sandbox.stub(stream, 'send');

        stream.write({
          time: '2017-05-06T16:48:19.763Z'
        });

        expect(sendStub.callCount).to.eql(1);
        const args = sendStub.getCall(0).args;
        expect(args).to.have.length(1);

        const payload = JSON.parse(args[0]);
        expect(payload).to.have.property('@timestamp', '2017-05-06T16:48:19.763Z');
        expect(payload).to.have.property('tags').eql(['tag1', 'tag2']);
        expect(payload).to.have.property('source', 'myServer.com/myApp');
        expect(payload).to.have.property('pid', 12345);
      });

      it('Should override the default message fields with the entry fields', () => {
        const stream = createStream({
          tags: ['tag1', 'tag2'],
          server: 'myServer.com',
          appName: 'myApp',
          pid: 12345
        });

        const sendStub = sandbox.stub(stream, 'send');

        stream.write({
          time: '2017-05-06T16:48:19.763Z',
          '@timestamp': 'overriden timestamp',
          msg: 'my message',
          tags: ['overriden tag'],
          source: 'overriden source',
          level: 'debug',
          myKey: 'myValue'
        });

        expect(sendStub.callCount).to.eql(1);
        const args = sendStub.getCall(0).args;
        expect(args).to.have.length(1);

        const payload = JSON.parse(args[0]);
        expect(payload).to.have.property('@timestamp', 'overriden timestamp');
        expect(payload).to.have.property('tags').eql(['overriden tag']);
        expect(payload).to.have.property('source', 'overriden source');
        expect(payload).to.have.property('myKey', 'myValue');
        expect(payload).to.have.property('pid', 12345);
        expect(payload).to.have.property('level', 'debug');
        expect(payload).to.have.property('message', 'my message');
      });

      it('Should accept a string containing valid JSON', () => {
        const stream = createStream({});

        const sendStub = sandbox.stub(stream, 'send');

        stream.write('{"time":"2017-05-06T16:48:19.763Z"}');

        expect(sendStub.callCount).to.eql(1);
        const args = sendStub.getCall(0).args;
        expect(args).to.have.length(1);
        const payload = JSON.parse(args[0]);
        expect(payload).to.have.property('@timestamp', '2017-05-06T16:48:19.763Z');
      });

      it('Should transform valid numeric levels into their string version', () => {
        const stream = createStream({});

        const sendStub = sandbox.stub(stream, 'send');

        stream.write({ time: '2017-05-06T16:48:19.763Z', level: 50 });

        expect(sendStub.callCount).to.eql(1);
        const args = sendStub.getCall(0).args;
        expect(args).to.have.length(1);
        const payload = JSON.parse(args[0]);
        expect(payload).to.have.property('level', 'error');
      });

      it('Should keep unknown numeric levels as is', () => {
        const stream = createStream({});

        const sendStub = sandbox.stub(stream, 'send');

        stream.write({ time: '2017-05-06T16:48:19.763Z', level: 51 });

        expect(sendStub.callCount).to.eql(1);
        const args = sendStub.getCall(0).args;
        expect(args).to.have.length(1);
        const payload = JSON.parse(args[0]);
        expect(payload).to.have.property('level', 51);
      });

      it('Should include string stream type in the message', () => {
        const stream = createStream({ type: 'alpha' });

        const sendStub = sandbox.stub(stream, 'send');

        stream.write({ time: '2017-05-06T16:48:19.763Z', level: 51 });

        expect(sendStub.callCount).to.eql(1);
        const args = sendStub.getCall(0).args;
        expect(args).to.have.length(1);
        const payload = JSON.parse(args[0]);
        expect(payload).to.have.property('type', 'alpha');
      });

      it('Should not include non string stream type in the message', () => {
        const stream = createStream({ type: 42 });

        const sendStub = sandbox.stub(stream, 'send');

        stream.write({ time: '2017-05-06T16:48:19.763Z', level: 51 });

        expect(sendStub.callCount).to.eql(1);
        const args = sendStub.getCall(0).args;
        expect(args).to.have.length(1);
        const payload = JSON.parse(args[0]);
        expect(payload).to.not.have.property('type');
      });

      it('Should not override stream pid with entry pid if any', () => {
        const stream = createStream({ pid: 42 });

        const sendStub = sandbox.stub(stream, 'send');

        stream.write({ time: '2017-05-06T16:48:19.763Z', pid: 43 });

        expect(sendStub.callCount).to.eql(1);
        const args = sendStub.getCall(0).args;
        expect(args).to.have.length(1);
        const payload = JSON.parse(args[0]);
        expect(payload).to.have.property('pid', 42);
      });
    });

    describe('connect', () => {
      it('Should create non tls socket when ssl_enable is false', () => {
        const tlsConnectSpy = sandbox.spy(tls, 'connect');

        createStream();

        expect(tlsConnectSpy.callCount).to.eql(0);
      });

      it('Should create tls socket when ssl_enable is true', () => {
        const tlsConnectStub = sandbox.stub(tls, 'connect').callsFake(() => new MockSocket());
        sandbox.stub(fs, 'readFileSync').callsFake(path => `content of ${path}`);

        const stream = createStream({
          ssl_enable: true,
          ssl_key: 'path/to/key',
          ssl_cert: 'path/to/cert',
          ssl_passphrase: 'passphrase',
          ca: ['path/to/ca1', 'path/to/ca2'],
          port: 12345,
          host: 'sslhost'
        });

        expect(tlsConnectStub.callCount).to.eql(1);
        const args = tlsConnectStub.getCall(0).args;
        expect(args).to.have.length(4);
        expect(args[0]).to.eql(12345);
        expect(args[1]).to.eql('sslhost');
        expect(args[2]).to.eql({
          key: 'content of path/to/key',
          cert: 'content of path/to/cert',
          passphrase: 'passphrase',
          ca: ['content of path/to/ca1', 'content of path/to/ca2']
        });
        const connectedCallback = args[3];

        expect(stream).to.have.property('connecting', true);
        expect(stream).to.have.property('connected', false);

        connectedCallback();

        expect(stream).to.have.property('connecting', false);
        expect(stream).to.have.deep.property('socket.encoding', 'UTF-8');
        expect(stream).to.have.property('connected', true);
      });

      it('Should hanlde socket errors', () => {
        sandbox.stub(tls, 'connect').callsFake(
          (port, host, options, callback) => {
            setTimeout(callback);
            return new MockSocket();
          }
        );
        const stream = createStream({ ssl_enable: true });
        const streamEmitStub = sandbox.stub(stream, 'emit');
        const socket = stream.socket;
        stream.socket.dispatchEvent('error', 'test error');
        expect(stream).to.have.property('connecting', false);
        expect(stream).to.have.property('connected', false);
        expect(socket).to.have.property('destroyCalled', true);
        expect(stream).to.have.property('socket', null);
        expect(streamEmitStub.withArgs('error', 'test error').callCount).to.equal(1);
      });

      it('Should hanlde socket timeouts', () => {
        sandbox.stub(tls, 'connect').callsFake(
          (port, host, options, callback) => {
            setTimeout(callback);
            return new MockSocket();
          }
        );
        const stream = createStream({ ssl_enable: true });
        const streamEmitStub = sandbox.stub(stream, 'emit');
        const socket = stream.socket;
        stream.socket.dispatchEvent('timeout');
        expect(socket).to.have.property('destroyCalled', true);
        expect(streamEmitStub.withArgs('timeout').callCount).to.equal(1);
      });

      it('Should hanlde socket timeouts with readyState still "open"', () => {
        sandbox.stub(tls, 'connect').callsFake(
          (port, host, options, callback) => {
            setTimeout(callback);
            return new MockSocket();
          }
        );
        const stream = createStream({ ssl_enable: true });
        stream.socket.readyState = 'open';
        const streamEmitStub = sandbox.stub(stream, 'emit');
        const socket = stream.socket;
        stream.socket.dispatchEvent('timeout');
        expect(socket).to.have.property('destroyCalled', false);
        expect(streamEmitStub.withArgs('timeout').callCount).to.equal(1);
      });

      it('Should hanlde connect event', () => {
        sandbox.stub(tls, 'connect').callsFake(
          (port, host, options, callback) => {
            setTimeout(callback);
            return new MockSocket();
          }
        );
        const stream = createStream({ ssl_enable: true });
        const streamEmitStub = sandbox.stub(stream, 'emit');
        stream.socket.dispatchEvent('connect');
        expect(stream).to.have.property('retries', 0);
        expect(streamEmitStub.withArgs('connect').callCount).to.equal(1);
      });

      it('Should hanlde close event when retries are still possible', done => {
        sandbox.stub(tls, 'connect').callsFake(
          (port, host, options, callback) => {
            setTimeout(callback);
            return new MockSocket();
          }
        );
        const stream = createStream({
          ssl_enable: true,
          max_connect_retries: 1,
          retry_interval: 1
        });

        const streamEmitStub = sandbox.stub(stream, 'emit');
        stream.connecting = false;
        stream.retries = 0;
        stream.socket.dispatchEvent('close');
        expect(streamEmitStub.withArgs('close').callCount).to.equal(1);
        expect(stream).to.not.have.property('silent');

        sandbox.stub(stream, 'connect').callsFake(done);
      });

      it('Should hanlde close event when retries are still possible but it\'s already connecting', done => {
        sandbox.stub(tls, 'connect').callsFake(
          (port, host, options, callback) => {
            setTimeout(callback);
            return new MockSocket();
          }
        );
        const stream = createStream({
          ssl_enable: true,
          max_connect_retries: 1,
          retry_interval: 1
        });

        const streamEmitStub = sandbox.stub(stream, 'emit');
        stream.connecting = true;
        stream.retries = 0;
        stream.socket.dispatchEvent('close');
        expect(streamEmitStub.withArgs('close').callCount).to.equal(1);
        expect(stream).to.not.have.property('silent');
        const connectStub = sandbox.stub(stream, 'connect');
        setTimeout(() => {
          expect(connectStub.callCount).to.equal(0);
          done();
        }, 10);
      });

      it('Should hanlde close event when no more retries are available', () => {
        sandbox.stub(tls, 'connect').callsFake(
          (port, host, options, callback) => {
            setTimeout(callback);
            return new MockSocket();
          }
        );
        const stream = createStream({
          ssl_enable: true,
          max_connect_retries: 1,
          retry_interval: 0
        });

        const streamEmitStub = sandbox.stub(stream, 'emit');
        stream.connecting = false;
        stream.retries = 1;

        const oldLogQueue = stream.log_queue;

        stream.socket.dispatchEvent('close');
        expect(streamEmitStub.withArgs('close').callCount).to.equal(1);

        expect(stream.log_queue !== oldLogQueue);
        expect(stream).to.have.property('silent', true);
      });
    });
    describe('flush', () => {
      it('Should send all messages and leave queue empty', () => {
        const stream = createStream();
        stream.log_queue.push({ message: 'a' });
        stream.log_queue.push({ message: 'b' });
        const sendStub = sandbox.stub(stream, 'sendLog');
        stream.flush();
        expect(sendStub.callCount).to.equal(2);
        expect(sendStub.withArgs('a').callCount).to.equal(1);
        expect(sendStub.withArgs('b').callCount).to.equal(1);
        expect(stream.log_queue).to.have.property('size', 0);
      });
    });
    describe('sendLog', () => {
      it('Should write log to the socket', () => {
        const stream = createStream();
        stream.sendLog('hello');
        expect(stream.socket.content).to.equal('hello\n');
      });
    });
    describe('send', () => {
      it('Should write log to the socket when connected', () => {
        const stream = createStream();
        stream.connected = true;
        const sendLogStub = sandbox.stub(stream, 'sendLog');
        stream.send('hello');
        expect(sendLogStub.callCount).to.equal(1);
        expect(sendLogStub.withArgs('hello').callCount).to.equal(1);
      });
    });
    describe('send', () => {
      it('Should store log in the queue when not connected', () => {
        const stream = createStream();
        stream.connected = false;
        const sendLogStub = sandbox.stub(stream, 'sendLog');
        stream.send('hello');
        expect(sendLogStub.callCount).to.equal(0);
        expect(stream.log_queue.pop()).to.eql({ message: 'hello' });
      });
    });
  });
});



src/lib/logstash.js
const bunyan = require('bunyan');
const net = require('net');
const fs = require('fs');
const os = require('os');
const tls = require('tls');
const CBuffer = require('CBuffer');
const _ = require('lodash');
const util = require('util');
const EventEmitter = require('events').EventEmitter;

const levels = new Map([
  [10, 'trace'],
  [20, 'debug'],
  [30, 'info'],
  [40, 'warn'],
  [50, 'error'],
  [60, 'fatal']
]);

/**
 * Creates a new instance of LogstashStream from the options.
 *
 * @param {objects} options The constructions options. See the constructor for details.
 * TODO: Improve this doc.
 *
 * @returns {LogstashStream} The bunyan stream that sends data to logstash
 */
function createLogstashStream(options) {
  return new LogstashStream(options);
}

/**
 * This class implements the bunyan stream contract with a stream that
 * sends data to logstash.
 *
 * @param {objects} options The constructions options. See the constructor for details.
 * TODO: Improve this doc.
 *
 * @constructor
 */
function LogstashStream(options) {
  EventEmitter.call(this);
  options = options || {};

  this.name = 'bunyan';
  this.level = options.level || 'info';
  this.server = options.server || os.hostname();
  this.host = options.host || '127.0.0.1';
  this.port = options.port || 9999;
  this.application = options.appName || process.title;
  this.pid = options.pid || process.pid;
  this.tags = options.tags || ['bunyan'];
  this.type = options.type;

  this.client = null;

  // ssl
  this.ssl_enable = options.ssl_enable || false;
  this.ssl_key = options.ssl_key || '';
  this.ssl_cert = options.ssl_cert || '';
  this.ca = options.ca || '';
  this.ssl_passphrase = options.ssl_passphrase || '';

  this.cbuffer_size = options.cbuffer_size || 10;

  // Connection state
  this.log_queue = new CBuffer(this.cbuffer_size);
  this.connected = false;
  this.socket = null;
  this.retries = -1;

  this.max_connect_retries = (typeof options.max_connect_retries === 'number') ? options.max_connect_retries : 4;
  this.retry_interval = options.retry_interval || 100;

  this.connect();
}
util.inherits(LogstashStream, EventEmitter);

/**
 * Writes a log entry to the steam.
 *
 * @param {object} entry The entry to write.
 * @returns {void}
 */
LogstashStream.prototype.write = function logstashWrite(entry) {
  let level;

  if (typeof (entry) === 'string') {
    entry = JSON.parse(entry);
  }

  const rec = _.cloneDeep(entry);

  level = rec.level;

  if (levels.has(level)) {
    level = levels.get(level);
  }

  const msg = {
    '@timestamp': new Date(rec.time).toISOString(),
    message: rec.msg,
    tags: this.tags,
    source: `${this.server}/${this.application}`,
    level
  };

  if (typeof (this.type) === 'string') {
    msg.type = this.type;
  }

  delete rec.time;
  delete rec.msg;

  // Remove internal bunyan fields that won't mean anything outside of
  // a bunyan context.
  delete rec.v;
  delete rec.level;

  rec.pid = this.pid;

  this.send(JSON.stringify(_.extend({}, msg, rec), bunyan.safeCycles()));
};

/**
 * Connects the stream to the remote logstash server specified in the options.
 *
 * @returns {void}
 */
LogstashStream.prototype.connect = function connect() {
  let options = {};
  const self = this;
  this.retries += 1;
  this.connecting = true;
  if (this.ssl_enable) {
    options = {
      key: this.ssl_key ? fs.readFileSync(this.ssl_key) : null,
      cert: this.ssl_cert ? fs.readFileSync(this.ssl_cert) : null,
      passphrase: this.ssl_passphrase ? this.ssl_passphrase : null,
      ca: this.ca ? this.ca.map(filePath => fs.readFileSync(filePath)) : null
    };
    this.socket = tls.connect(this.port, this.host, options, () => {
      self.socket.setEncoding('UTF-8');
      self.announce();
      self.connecting = false;
    });
  } else {
    this.socket = new net.Socket();
  }
  this.socket.unref();

  this.socket.on('error', (err) => {
    self.connecting = false;
    self.connected = false;
    self.socket.destroy();
    self.socket = null;
    self.emit('error', err);
  });

  this.socket.on('timeout', () => {
    if (self.socket.readyState !== 'open') {
      self.socket.destroy();
    }
    self.emit('timeout');
  });

  this.socket.on('connect', () => {
    self.retries = 0;
    self.emit('connect');
  });

  this.socket.on('close', () => {
    self.connected = false;

    if (self.max_connect_retries < 0 || self.retries < self.max_connect_retries) {
      if (!self.connecting) {
        setTimeout(() => {
          self.connect();
        }, self.retry_interval).unref();
      }
    } else {
      self.log_queue = new CBuffer(self.cbuffer_size);
      self.silent = true;
    }
    self.emit('close');
  });

  if (!this.ssl_enable) {
    this.socket.connect(self.port, self.host, () => {
      self.announce();
      self.connecting = false;
    });
  }
};

/**
 * Announces that the stream is connected. Will flush any messages in the queue.
 *
 * @returns {void}
 */
LogstashStream.prototype.announce = function announce() {
  const self = this;
  self.connected = true;
  self.flush();
};

/**
 * Flushes the queue, sending all messages that have not been sent yet to the remote
 * destination.
 *
 * @returns {void}
 */
LogstashStream.prototype.flush = function flush() {
  const self = this;

  let message = self.log_queue.pop();
  while (message) {
    self.sendLog(message.message);
    message = self.log_queue.pop();
  }

  self.log_queue.empty();
};

/**
 * Immediately writes a string to the undelying socket.
 *
 * @param {string} message The string to write.
 * @returns {void}
 */
LogstashStream.prototype.sendLog = function sendLog(message) {
  this.socket.write(`${message}\n`);
};


/**
 * Sends a string message. The message will be immediately sent if the stream
 * is already connected, or queued if the stream is not connected yet.
 * @param {string} message The string to send
 * @returns {void}
 */
LogstashStream.prototype.send = function logstashSend(message) {
  const self = this;

  // send tcp logs
  if (!self.connected) {
    self.log_queue.push({
      message
    });
  } else {
    self.sendLog(message);
  }
};

module.exports = {
  createStream: createLogstashStream,
  LogstashStream
};
